src/basic/mkdir.c

```diff
@@ -194,56 +195,16 @@ int mkdir_p_safe(const char *prefix, const char *path, mode_t mode, uid_t uid, g
         return mkdir_p_internal(prefix, path, mode, uid, gid, flags, mkdirat_errno_wrapper);
 }
 
-int mkdir_p_root(const char *root, const char *p, uid_t uid, gid_t gid, mode_t m) {
-        _cleanup_free_ char *pp = NULL, *bn = NULL;
-        _cleanup_close_ int dfd = -1;
-        int r;
-
-        r = path_extract_directory(p, &pp);
-        if (r == -EDESTADDRREQ) {
-                /* only fname is passed, no prefix to operate on */
-                dfd = open(".", O_RDONLY|O_CLOEXEC|O_DIRECTORY);
-                if (dfd < 0)
-                        return -errno;
-        } else if (r == -EADDRNOTAVAIL)
-                /* only root dir or "." was passed, i.e. there is no parent to extract, in that case there's nothing to do. */
-                return 0;
-        else if (r < 0)
-                return r;
-        else {
-                /* Extracting the parent dir worked, hence we aren't top-level? Recurse up first. */
-                r = mkdir_p_root(root, pp, uid, gid, m);
-                if (r < 0)
-                        return r;
-
-                dfd = chase_symlinks_and_open(pp, root, CHASE_PREFIX_ROOT, O_RDONLY|O_CLOEXEC|O_DIRECTORY, NULL);
-                if (dfd < 0)
-                        return dfd;
-        }
+int mkdir_p_root(const char *root, const char *path, mode_t mode, uid_t uid, gid_t gid, MkdirFlags flags) {
+        _cleanup_free_ char *p = NULL;
 
-        r = path_extract_filename(p, &bn);
-        if (r == -EADDRNOTAVAIL) /* Already top-level */
-                return 0;
-        if (r < 0)
-                return r;
-
-        if (mkdirat(dfd, bn, m) < 0) {
-                if (errno == EEXIST)
-                        return 0;
-
-                return -errno;
-        }
-
-        if (uid_is_valid(uid) || gid_is_valid(gid)) {
-                _cleanup_close_ int nfd = -1;
-
-                nfd = openat(dfd, bn, O_RDONLY|O_CLOEXEC|O_DIRECTORY|O_NOFOLLOW);
-                if (nfd < 0)
-                        return -errno;
+        assert(path);
 
-                if (fchown(nfd, uid, gid) < 0)
-                        return -errno;
+        if (root) {
+                p = path_join(root, path);
```

Comment:
mkdir_p_root() was about chase_symlinks(), i.e. that we mkdir a dir and don't leave the specified root while resolving the prefix to that dir. But here you just joine the path and then mkdir_p_safe() basically invokes mkdir() on the result, without understanding of the root parth thing